src/sqlfluff/core/linter/linter.py

```diff
@@ -587,6 +588,8 @@ def parse_string(
         # Scan the raw file for config commands.
         config.process_raw_file_for_config(in_str)
         rendered = self.render_string(in_str, fname, config, encoding)
+        if templated_file is not None:
+            templated_file[:] = [rendered.templated_file]
```

Comment:
Python parameters are always "pass by value". So to implement "output parameters", you have to pass in a collection and update the collection instead. If we did this, it would assign `templated_file` to point to a new, different list, _not_ updating the one passed in by the caller:
```
templated_file = [rendered.templated_file]
```

https://www.learnbyexample.org/python-list-slicing/#modify-multiple-list-values

I considered updating the function to return two values rather than use this trickery, but it's a public `Linter` function, so I thought it inappropriate to change the API to return something very few people care about.

Reply:
Oh I see now. And that is why we're passing in `Optional[List[TemplatedFile]]` rather than just `Optional[TemplatedFile]` I presume?

But not sure I like it to be honest. Seems a bit dishonest to pass input parameters and then alter them with this trickery. At the very least should add a comment to explain this.

However I can also understand hesitancy to alter `parse_string` (even though technically you have already in this PR, with the optional input param - but at least that's optional). What do you think about creating a `parse_string_and_template` function (perhaps with a better name!) that does this properly by returning two values?

Then `parse_string` could either be left along (there's not much code to duplicate to be honest!) or, perhaps better, it could call the `parse_string_and_template` function with blank template and strip out the returned template, so the API of `parse_string` remains the same?

Reply:
I'm a little hesitant about adding another public function to this user-facing class. I'll see if there's a way to add it as a standalone function. It'll still be "public", but not "staring users in the face", if that makes sense.

Reply:
Actually, the `Linter` class already has a large number of "public" functions that are intended mainly for internal use, so I'm okay adding another function there.